### PR TITLE
Feature/add coreset method targeting CMMD

### PR DESCRIPTION
**Pull request by db091756**
### PR Type

- Feature
- Tests

### Description

- Adds CMMD metric
- Adds GreedyCMMD, an algorithm to target CMMD
- Adds util functions to sample batch indices, invert arrays
- Adds simple example

### How Has This Been Tested?

- Added integration tests for example
- Added unit tests for CMMD metric and coreset algorithm
- Added unit tests for util functions

### Does this PR introduce a breaking change?

NA

### Screenshots

NA

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: coreax/util.py [hunk 1]
```diff
@@ -193,6 +194,185 @@ def solve_qp(kernel_mm: ArrayLike, kernel_matrix_row_sum_mean: ArrayLike) -> Arr
     return sol.primal
 
 
+@partial(jit, static_argnames=("oversampling_parameter", "power_iterations"))
+def randomised_eigendecomposition(
```
Comment (tc85324) [inline coreax/util.py:198]: Please can this be separated out into another pull/issue.

File: coreax/util.py [hunk 2]
```diff
@@ -193,6 +194,185 @@ def solve_qp(kernel_mm: ArrayLike, kernel_matrix_row_sum_mean: ArrayLike) -> Arr
...
+    # the eigenvectors.
+    array_approximation = q.T @ array @ q
+    array_approximate_eigenvalues, eigenvectors = jnp.linalg.eigh(array_approximation)
+    array_approximate_eigenvectors = q @ eigenvectors
+
+    return array_approximate_eigenvalues, array_approximate_eigenvectors
+
+
+@partial(jit, static_argnames=("regularisation_parameter", "rcond"))
+def invert_regularised_array(
```
Comment (tc85324) [inline coreax/util.py:256]: Does the `identity` need to be explictily provided, or can it be compute from the passed array?

File: coreax/metrics.py [hunk 3]
```diff
@@ -428,3 +429,200 @@ def sum_weighted_pairwise_distances(
...
+            regularisation_parameter=self.regularisation_parameters[1].item(),
+            identity=identity_2,
+        )
+
+        # Compute each term in the CMMD
+        term_1 = (
+            inverse_feature_gramian_1
+            @ response_gramian_1
+            @ inverse_feature_gramian_1
+            @ feature_gramian_1
```
Comment (tc85324) [inline coreax/metrics.py:595]: Does the `inverse_feature @ feature` not simplify to the indentity matrix, or is this becuase we aren't computing a real inverse, but a regularized psuedo-inverse?

File: coreax/metrics.py [hunk 4]
```diff
@@ -428,3 +429,200 @@ def sum_weighted_pairwise_distances(
...
+        )
+
+        # Compute CMMD
+        squared_result = jnp.trace(term_1) + jnp.trace(term_2) - 2 * jnp.trace(term_3)
+        if squared_result < 0:
+            warn(
+                f"Squared CMMD ({round(squared_result.item(), 4)}) is negative,"
+                + " increase precision threshold or regularisation strength.",
+                Warning,
+                stacklevel=1,
```
Comment (tc85324) [inline coreax/metrics.py:617]: Stacklevel probably wants to be 2. We have another example somewhere in the code where this same change should be made.

**Review by tc85324 — COMMENTED** [review]
I've had a quick look through the code; the current size of this pull makes it challenging to review so I've suggested some places where parts can be separated out. Once the size has been reduced we can take a better look at functionality/correctenss/efficiency etc...

File: coreax/refine.py [hunk 5]
```diff
@@ -744,9 +744,248 @@ def _no_change_rev(
         return jnp.asarray(coreset_indices)
 
 
+class RefineCMMD(Refine):
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/refine.py:747]: Will do!

File: coreax/metrics.py [hunk 6]
```diff
@@ -428,3 +429,200 @@ def sum_weighted_pairwise_distances(
...
+        assert block_size is None, "CMMD computation does not support blocking"
+        assert weights_x is None, "CMMD computation does not support weights"
+        assert weights_y is None, "CMMD computation does not support weights"
+
+        return self.conditional_maximum_mean_discrepancy(dataset_1, dataset_2)
+
+    # pylint: enable=arguments-renamed
+    # Clarity means I need to assign more local variables
+    # pylint: disable=too-many-locals
+    def conditional_maximum_mean_discrepancy(
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:526]: Agree, I used this structure to match the MMD structure where they separate into diff function calls based on user supplied weights/blocking. The argument for leaving it this way is that I want to add methods to approximate the CMMD using that randomised eigen decomposition util function and so I would likely be re-adding it at that point

File: coreax/util.py (hunk 2 above)
Comment (db091756) [reply coreax/util.py:256]: I wrote it this way for two reasons:

1. I figured it was more efficient to supply the identity array rather than recreating a new one each time within the function as this would add overhead (I think) to the vmapped version of this function
2. Because the main point of the function is to let us invert a growing array while keeping shapes static. The "identity" array changes at each iteration within GreedyCMMD; it starts as a an array of zeros with a 1 in the top left element, then at each iteration we add 1 to the next element on the diagonal to make sure we are only "inverting" the correct subarray. Hopefully this makes sense. I am not actually sure if just inputting the regular full-size identity matrix here would work or not, I suspect not as then the least square function would be trying to solve a least-squares problem Ax=b where A has rows of zeros corresponding to an element of b that is 1. I likely checked this but I'm not 100% sure

File: coreax/metrics.py (hunk 3 above)
Comment (db091756) [reply coreax/metrics.py:595]: Yeah because it is a regularised inverse this doesn't just become the identity matrix

File: coreax/metrics.py (hunk 4 above)
Comment (db091756) [reply coreax/metrics.py:617]: Will change!

File: coreax/metrics.py [hunk 7]
```diff
@@ -428,3 +429,200 @@ def sum_weighted_pairwise_distances(
...
+            concatenated after the features
+        :param block_size: :data:`None`, included for compatibility reasons
+        :param weights_x: :data:`None`, included for compatibility reasons
+        :param weights_y: :data:`None`, included for compatibility reasons
+        :return: Conditional maximum mean discrepancy as a 0-dimensional array
+        """
+        # Make sure that compatibility params are None
+        assert block_size is None, "CMMD computation does not support blocking"
+        assert weights_x is None, "CMMD computation does not support weights"
+        assert weights_y is None, "CMMD computation does not support weights"
```
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/metrics.py:519]: A refactor of the `Metric` abstract might be something for future discussion, so I would encourage raising a feature request ticket if you think it would be useful. This might be covered if we introduced weighted datastructures/arrays.

[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:519]: Yeah I think this is something that will make sense to change in the future, I'll set up an issue for it. Should I remove the assert statements for now?

File: coreax/util.py (hunk 2 above)
Comment (db091756) [reply coreax/util.py:198]: Had another look at this and it seems like 2. can be ignored, my worry with the least squares problem failing doesn't seem to be happening. Probably still fine to keep as is for the sake of 1.

File: .cspell/custom_misc.txt [hunk 8]
```diff
@@ -3,26 +3,33 @@ approximator
 approximators
 archiveprefix
 cand
+cmes
```
[the remark this answers is not in the document]
Comment (db091756) [reply .cspell/custom_misc.txt:6]: conditional mean embedding(s)

File: coreax/util.py [hunk 9]
```diff
@@ -210,6 +211,128 @@ def solve_qp(kernel_mm: ArrayLike, kernel_matrix_row_sum_mean: ArrayLike) -> Arr
...
+        partial(
+            invert_regularised_array,
+            regularisation_parameter=regularisation_parameter,
+            identity=identity,
+            rcond=rcond,
+        )
+    )(stacked_arrays)
+
+
+def sample_batch_indices(
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/util.py:301]: I expect that this function will be used in at least 2 more coreset algorithms I will be writing after this one

File: coreax/coresubset.py [hunk 10]
```diff
@@ -406,7 +407,7 @@ def fit_to_size(self, coreset_size: int) -> None:
 
         # Assign coreset indices and coreset to the object
         self.coreset_indices = random_indices
-        self.coreset = self.original_data.pre_coreset_array[random_indices]
+        self.coreset = self.original_data.pre_coreset_array[random_indices, :]
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/coresubset.py:410]: Not sure when I did this, removed

File: coreax/coresubset.py [hunk 11]
```diff
@@ -617,3 +618,318 @@ def _loop_body(
...
+
+        # Compute original feature and response kernel gramians.
+        x = self.original_data.pre_coreset_array[:, : self.num_feature_dimensions]
+        y = self.original_data.pre_coreset_array[:, self.num_feature_dimensions :]
+        num_data_pairs = self.original_data.pre_coreset_array.shape[0]
+
+        feature_gramian = self.feature_kernel.compute(x, x)
+        response_gramian = self.response_kernel.compute(y, y)
+
+        # Invert the feature gramian
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/coresubset.py:767]: Agreed, done

File: coreax/coresubset.py [hunk 12]
```diff
@@ -617,3 +618,318 @@ def _loop_body(
...
+        self,
+        random_key: coreax.util.KeyArrayLike,
+        *,
+        feature_kernel: coreax.kernel.Kernel,
+        response_kernel: coreax.kernel.Kernel,
+        num_feature_dimensions: int,
+        regularisation_parameter: float = 1e-6,
+        unique: bool = True,
+        batch_size: int | None = None,
+        refine_method: coreax.refine.RefineCMMD | None = None,
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/coresubset.py:684]: Fixed

File: .cspell/library_terms.txt [hunk 13]
```diff
@@ -26,19 +26,22 @@ dispname
 dollarmath
 dtype
 dtypes
+eigh
```
[the remark this answers is not in the document]
Comment (db091756) [reply .cspell/library_terms.txt:29]: Fixed

File: .cspell/custom_misc.txt [hunk 14]
```diff
@@ -3,26 +3,33 @@ approximator
...
 GCHQ
+gramian
+gramians
 kbar
 kdtree
 kernelised
 kernelized
 ml.p3.8xlarge
 nonnegative
+orthonormalised
```
[the remark this answers is not in the document]
Comment (db091756) [reply .cspell/custom_misc.txt:26]: Fixed both

File: coreax/coresubset.py [hunk 15]
```diff
@@ -617,3 +618,318 @@ def _loop_body(
...
+        )
+        aux_data = {
+            "num_feature_dimensions": self.num_feature_dimensions,
+            "regularisation_parameter": self.regularisation_parameter,
+            "unique": self.unique,
+            "batch_size": self.batch_size,
+            "refine_method": self.refine_method,
+            "unused_kernel": self.kernel,
+            "unused_weights_optimiser": self.weights_optimiser,
+            "unused_kernel_mean": self.kernel_matrix_row_sum_mean,
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/coresubset.py:740]: Will change key name for now

File: coreax/coresubset.py [hunk 16]
```diff
@@ -617,3 +618,318 @@ def _loop_body(
...
+        coreset_indices = -1 * jnp.ones(coreset_size, dtype=jnp.int32)
+
+        # Initialise a zeros matrix that will eventually become a coreset_size x
+        # coreset_size identity matrix as we iterate to the full coreset size. This will
+        # allow us to invert static matrices using coreax.util.invert_regularised_array.
+        coreset_identity = jnp.zeros((coreset_size, coreset_size))
+
+        # body_fn of GreedyCMMD and RefineCMMD are very similar, might be worth
+        # thinking of how this could be avoided?
+        # pylint: disable=duplicate-code
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/coresubset.py:824]: Removed

File: coreax/coresubset.py [hunk 17]
```diff
@@ -617,3 +618,318 @@ def _loop_body(
...
+        @jit
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/coresubset.py:825]: Removed

File: coreax/coresubset.py [hunk 18]
```diff
@@ -617,3 +618,318 @@ def _loop_body(
...
+            index_to_include_in_coreset = all_possible_next_coreset_indices[
+                loss.argmin(), i
+            ]
+
+            # Repeat the chosen coreset index into the ith column of the array of
+            # possible next coreset indices and replace the (i+1)th column with the next
+            # batch of possible coreset indices.
+            all_possible_next_coreset_indices = all_possible_next_coreset_indices.at[
+                :, [i, i + 1]
+            ].set(
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/coresubset.py:889]: Fixed

File: coreax/coresubset.py [hunk 19]
```diff
@@ -617,3 +618,318 @@ def _loop_body(
...
+        # body_fn of GreedyCMMD and RefineCMMD are very similar, might be worth
+        # thinking of how this could be avoided?
+        # pylint: disable=duplicate-code
+        @jit
+        def _greedy_body(
+            i: int,
+            val: tuple[ArrayLike, ArrayLike, ArrayLike],
+        ) -> tuple[ArrayLike, ArrayLike, ArrayLike]:
+            r"""Execute main loop of GreedyCMMD."""
+            # Unpack the components of the loop variables
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/coresubset.py:831]: Removed more obvious ones

File: coreax/coresubset.py [hunk 20]
```diff
@@ -617,3 +618,318 @@ def _loop_body(
...
+            return (
+                current_coreset_indices,
+                current_identity,
+                all_possible_next_coreset_indices,
+            )
+
+        # pylint: enable=duplicate-code
+
+        # Greedily select coreset points
+        (coreset_indices, _, _) = lax.fori_loop(
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/coresubset.py:913]: Fixed

File: coreax/metrics.py [hunk 21]
```diff
@@ -428,3 +429,195 @@ def sum_weighted_pairwise_distances(
...
+        self.precision_threshold = precision_threshold
+
+        # Initialise parent
+        super().__init__()
+
+    # Disable pylint warning for arguments-renamed as the use of x and y as arguments
+    # for different datasets is bound to create misunderstandings when we are
+    # considering supervised data.
+    # pylint: disable=arguments-renamed
+    def compute(
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:492]: Agreed, removed

File: coreax/metrics.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:521]: Will change invert regularised array util function to do approximate inversion

File: coreax/metrics.py [hunk 22]
```diff
@@ -428,3 +429,195 @@ def sum_weighted_pairwise_distances(
...
+            \{(x_i, y_i)\}_{i=1}^n` of ``n`` pairs with :math:`x\in\mathbb{R}^d` and
+            :math:`y\in\mathbb{R}^p`, responses should be concatenated after the
+            features
+        :param dataset_2: Dataset :math:`\mathcal{D}^{(2)} = \{(\tilde{x}_i,
+            \tilde{y}_i)\}_{i=1}^n` of ``m`` pairs with :math:`\tilde{x}\in\mathbb{R}^d`
+            and :math:`\tilde{y}\in\mathbb{R}^p`, responses should be
+            concatenated after the features
+        """
+        # Extract and format features and responses from D1 and D2
+        x1 = jnp.atleast_2d(dataset_1[:, : self.num_feature_dimensions])
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:555]: Removed some unnecessary local vars

File: coreax/metrics.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:492]: Done
